components/eam/src/chemistry/modal_aero/modal_aero_amicphys.F90

```diff
@@ -3584,334 +3576,608 @@ end subroutine mam_soaexch_1subarea
+
+      subroutine compute_size_factor(imode, size_factor)
+        ! Compute size factor for a mode
+        implicit none
+
+        integer,  intent(in) :: imode     !mode number
+        real(r8), intent(inout) :: size_factor(:) !size factor
+
+        size_factor(imode) = (pi/6.)*exp(4.5*(alnsg_aer(imode)**2))
+
+      end subroutine compute_size_factor
+
+      !----------------------------------------------------------------------
+      !----------------------------------------------------------------------
+
+      pure function compute_vol_to_num_ratio(imode, diameter) result(v2n)
+        !compute volume to number ratio for a mode
+        implicit none
+        integer,  intent(in) :: imode
+        real(r8), intent(in) :: diameter(:) !particle diameter [m]
+
+        real(r8) :: v2n !return value
+
+        v2n = ( 1._r8 / ( (pi/6._r8)* &
+             (diameter(imode)**3._r8)*exp(4.5_r8*alnsg_aer(imode)**2._r8) ) )
+
+      end function compute_vol_to_num_ratio
+
+      !----------------------------------------------------------------------
+      !----------------------------------------------------------------------
+
+      subroutine compute_dryvol_change_in_src_mode(nmode, nspec, dest_mode_of_mode, &
+           iscldy, qi_vmr, qi_del_growth, qcld_vmr, qcld_del_growth, &
+           dryvol_a, deldryvol_a, dryvol_c, deldryvol_c)
+
+        integer,  intent(in):: nmode ! total number of modes
+        integer,  intent(in):: nspec !total number of species in a mode
+        integer,  intent(in):: dest_mode_of_mode(:) ! destination mode for a mode
+
+        logical,  intent(in) :: iscldy ! true if it is a cloudy cell
+
+        real(r8), intent(in) :: qi_vmr(:,:)           ! mass mixing ratios (mmr) [kmol/kmol]
+        real(r8), intent(in) :: qi_del_growth(:,:) !growth in mmr [kmol/kmol]
+
+        real(r8), intent(in), optional :: qcld_vmr(:,:)
+        real(r8), intent(in), optional :: qcld_del_growth(:,:)
+
+        !intent-outs
+        real(r8), intent(out) :: dryvol_a(:), dryvol_c(:)       !dry volumes (before growth) [m3/kmol-air]
+        real(r8), intent(out) :: deldryvol_a(:), deldryvol_c(:) !change in dry volumes [m3/kmol-air]
+
+        integer :: imode
+        integer :: dest_mode
+
+        !For each mode, compute the initial (before growth) dryvolume and the growth in dryvolume
+        do imode = 1, nmode
+           !compute dry volume only for modes participating in inter-modal transfer
+           dest_mode = dest_mode_of_mode(imode)
+           if (dest_mode <= 0) cycle
+
+           !compute dry volumes (before growth) and its change for interstitial aerosols
+           call dryvolume_change(imode, nspec, qi_vmr, qi_del_growth, & !input
+                dryvol_a(imode), deldryvol_a(imode)) !output
+
+           if ( iscldy ) then ! if this grid cell has cloud
+              !if a grid cell is cloudy, clloud borne quantities has to be present
```

Comment:
The logical parameter `iscldy` determines whether the optional parameters `qcld_vmr` and `qcld_del_growth` should be present. This suggests to me that this subroutine should probably be split into cloudy and interstitial versions that accept different sets of parameters.